Mixed Swift/C++ targets support

dblsaiko opened the pull request:
Closes #13203.

Makes Swift target generation use the main code path. This allows mixing Swift and C++ sources inside of a single target. (All Swift sources are still compiled with one swiftc invocation since [per-file compilation is not supported](https://github.com/swiftlang/sourcekit-lsp/issues/1980#issuecomment-2660422475).)

This also adds support for shared library Swift targets.

Also creates a Ninja target for a C++/Obj-C exported header, which those languages can use to call Swift code. This header is automatically added as a dependency to targets that can use it.

~~Caveat: using this to call Swift code from C++ requires manually setting CXX to the Swift-bundled clang++ (on non-macOS), since it requires special support in the compiler which the normal clang doesn't come with. Meson would automatically have to switch to using the Swift-bundled clang when a project uses both Swift and C/C++/Obj-C. Not sure if that is desirable or even doable. Could maybe throw an error instead when the mismatch is detected.~~ After thinking about it some more, CXX should just be set correctly by the user for this.

To do:
- [x] Test on macOS, specifically with Obj-C
- [x] Version checks (C++ bridging is available in 5.9 and above)
- [x] Test cases
- Documentation (#15575)

```meson
# Linking Swift lib with C++ executable
swiftlib = static_library('swiftlib', 'swiftlib.swift', swift_args: ['-parse-as-library'])
cpp_exe = executable('cpp_exe', 'main.cpp', link_with: [swiftlib], include_directories: [swiftlib.private_dir_include()])

# Single Swift/C++ target
combined_swift_cpp = executable('combined', 'combined.swift', 'combined.h', 'combined.cpp')
```

dblsaiko committed:
Support for generating C-family header from Swift code

Allows calling Swift code from C-family languages by adding the Swift
library's private dir to the C-family target's include directories.

dblsaiko committed:
Correctly link shared Swift library

dblsaiko committed:
Use normal compile target generation code path for Swift

This allows targets containing both C and Swift sources, and allows
linking a C++ target against a Swift static library by using the
Swift compiler as a linker.

thewildtree commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -1004,6 +1001,26 @@ def generate_target(self, target) -> None:
+
+        # For linking with a Swift target in a Obj-C/C++ target, include the Swift target's generated header.
+        if any({lang in target.compilers for lang in ['objc', 'cpp', 'objcpp']}):
+            for header in self.determine_swift_dep_headers(target):
```
This unfortunately breaks if said target links to a part of a custom target:

```
Traceback (most recent call last):
  File "[...]/meson/mesonbuild/mesonmain.py", line 193, in run
    return options.run_func(options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "[...]/meson/mesonbuild/msetup.py", line 365, in run
    app.generate()
  File "[...]/meson/mesonbuild/msetup.py", line 188, in generate
    return self._generate(env, capture, vslite_ctx)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "[...]/meson/mesonbuild/msetup.py", line 253, in _generate
    captured_compile_args = intr.backend.generate(capture, vslite_ctx)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "[...]/meson/mesonbuild/backend/ninjabackend.py", line 663, in generate
    self.generate_target(t)
  File "[...]/meson/mesonbuild/backend/ninjabackend.py", line 1021, in generate_target
    for header in self.determine_swift_dep_headers(target):
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "[...]/meson/mesonbuild/backend/ninjabackend.py", line 2205, in determine_swift_dep_headers
    if self.is_swift_target(l):
       ^^^^^^^^^^^^^^^^^^^^^^^
  File "[...]/meson/mesonbuild/backend/backends.py", line 516, in is_swift_target
    for s in target.sources:
             ^^^^^^^^^^^^^^
AttributeError: 'CustomTargetIndex' object has no attribute 'sources'
```

Something like this seems to help:

```diff
diff --git a/mesonbuild/backend/ninjabackend.py b/mesonbuild/backend/ninjabackend.py
index 2dda038ea..dc60a60ad 100644
--- a/mesonbuild/backend/ninjabackend.py
+++ b/mesonbuild/backend/ninjabackend.py
@@ -2202,7 +2202,7 @@ class NinjaBackend(backends.Backend):
     def determine_swift_dep_headers(self, target):
         result = []
         for l in target.link_targets:
-            if self.is_swift_target(l):
+            if isinstance(l, build.BuildTarget) and self.is_swift_target(l):
                 result.append(self.swift_generated_header_file_name(l))
         return result
```

2xsaiko replied:
Ah, thanks! That check should probably go into is_swift_target tbh (unless that already only accepts a BuildTarget per the parameter types).

2xsaiko replied:
I added it in is_swift_target after all and relaxed the type bounds since that seems less error-prone :)

Also, looks like mypy doesn't complain about calling is_swift_target with an invalid type even when are type annotations. Smh

thewildtree commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2169,6 +2186,10 @@ def swift_module_file_name(self, target):
         return os.path.join(self.get_target_private_dir(target),
                             self.target_swift_modulename(target) + '.swiftmodule')
 
+    def swift_generated_header_file_name(self, target):
+        return os.path.join(self.get_target_private_dir(target),
+                            self.target_swift_modulename(target) + '-Swift.h')
```
with `-working-directory` being set, this and the module path above seem to need an absolute path:

```suggestion
    def swift_generated_header_file_name(self, target):
        return os.path.join(self.get_target_private_dir_abs(target),
                            self.target_swift_modulename(target) + '-Swift.h')
```

2xsaiko replied:
Does it? Works fine for me. These are used in the Ninja file where they're supposed to be relative to the top build dir, and in case of the header generation that isn't run in the private target dir but instead the top build dir.

thewildtree replied:
You're right of course, I had leftover local changes which affected this, so please ignore. Sorry for the noise.

2xsaiko commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2269,26 +2297,30 @@ def generate_swift_target(self, target) -> None:
+        if not isinstance(target, build.Executable):
+            elem = NinjaBuildElement(self.all_outputs, out_module_name, rulename, abssrc)
+            elem.add_dep(in_module_files + rel_generated)
+            elem.add_item('ARGS', compile_args + abs_generated + module_includes + swiftc.get_mod_gen_args())
```
I wanted to add this a couple times now because I keep seeing the mismatch between the args of the two targets -- but as far as I understand, using bridging headers in a library seems to be not supported: https://forums.swift.org/t/spm-library-with-bridging-header/4945

Not sure if that was the original intention of leaving this away or if it was forgotten (I didn't write these two arg lists)

It seems to be allowed for frameworks (https://developer.apple.com/documentation/swift/importing-objective-c-into-swift), so I assume it's related to the install interface. Packaging a Swift library would involve installing both the Swift module files but additionally the C header, and the information about which libraries to link, for client Swift libraries to be able to import it. I assume there's no defined interface for that.

I'd rather add a failure check for this. You should instead use a module map to create a C module that you can import from Swift. Here an example for the matio library:

CMatio/meson.build:
```meson
matio_dep = dependency('matio')

CMatio_dep = declare_dependency(
  include_directories: ['.'],
  dependencies: [matio_dep],
)
```

CMatio/matio.h:
```c
#include_next <matio.h>
```

CMatio/module.modulemap:
```
module CMatio [extern_c] {
  header "matio.h"
}
```

Then use CMatio_dep in your Swift library target.

thewildtree replied:
I see, you're right, I never noticed it was already missing before your changes. I was curious because my build, which relies on a header being included like that, works perfectly fine without this PR.

Turns out previously the .swiftmodule just wasn't being used, so it was never actually built I assume, hence no errors.

I wasn't aware this wasn't meant to be supported at all, in which case I agree blocking it - like Xcode does - sounds best.

dblsaiko committed:
Fix crash linking Swift target with custom target

dblsaiko committed:
Verify that C++ can import Swift headers

dblsaiko committed:
swift: Add color diagnostics args

dblsaiko committed:
swift: Fix up various compiler flags

ADKaster commented on mesonbuild/compilers/swift.py:
```diff
@@ -99,18 +105,37 @@ def get_output_args(self, target: str) -> T.List[str]:
     def get_mod_gen_args(self) -> T.List[str]:
         return ['-emit-module']
 
+    def get_header_gen_args(self, header_name: str) -> T.List[str]:
+        return ['-parse', '-emit-objc-header', '-emit-objc-header-path', header_name]
```
This is probably out of scope for this PR, but swift 6's C++ interoperability features introduce a new way to do interop that doesn't use an objective C header.

Example: https://github.com/swiftlang/swift-cmake-examples/blob/36f0a9402d6e7a6225553c628f1b3bc6a78d33db/3_bidirectional_cxx_interop/cmake/modules/AddSwift.cmake#L16

Docs: https://www.swift.org/documentation/cxx-interop/#exposing-swift-apis-to-c

It works something like this:

```
swiftc -typecheck \
      -I<some_cxx_path> -I<some_swift_path> -I<swift-intrinsic-search-paths> \
      <all the swift source files for SomeModule>.swift
      -module-name SomeModule \
      -cxx-interoperability-mode=default \
      -emit-clang-header-path "$targetBuildDir/include/SomeModule-Export.h"
```

by `#include "SomeModule-Export.h"`, you can call into swift code directly from C++.

I would expect for "new" projects, using this approach will be preferred to the old objective C bridging header approach.

dblsaiko replied:
This was already written for Swift 6 on Linux and macOS using the new C++ interop. As far as I can tell, -emit-clang-header-path and -emit-objc-header-path do the same thing. I kept the latter for consistency with -emit-objc-header, which also exports a C++ header when C++ interop is enabled. Which I want here to explicitly tell the compiler to generate a header from the swiftmodule file (I'm not using the sources because the swiftmodule is enough and this also allows me to generate a header for installed libraries which I plan on doing)

I could switch to -typecheck and
-emit-clang-header-path though, not sure if -typecheck does anything extra.

Hmm, actually, now that I think about it, how does Apple handle this, do they ship the C header with a Swift library, or do they let the consumer of the library generate a header as needed? (I'd prefer the latter personally because it's less redundant stuff to install.)

EDIT: For future reference, it is the former. The C header is installed as part of the interface and not generated on import.

2xsaiko commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2263,6 +2263,10 @@ def generate_swift_target(self, target) -> None:
+        if header_imports != []:
+            # Swift writes a PCH file for bridging headers. Without this flag, it puts it somewhere in /tmp.
+            # '.' is already the target private dir.
+            header_imports += swiftc.get_pch_output_dir_args('.')
```
Looking at this again, I don't know if we actually want this. Might be better to leave these in the temp directory where they will be automatically cleaned up eventually, rather than the build directory where that is not the case until it is deleted. And just hope this is solved in a less stupid way by Swift in the future (such as not writing a PCH file to disk in the first place)

2xsaiko replied:
Yeah. Away it goes.

dblsaiko committed:
swift: Combine header and swiftmodule target into one interface target

Since Swift 6.3, the previous way I generated the Swift export header
(with -parse and passing the .swiftmodule) no longer works. Since
this was never an intentional behavior of the Swift compiler, re-do
the target model here to have one target for the object files and one
target to produce the Swift interface (i.e. .swiftmodule, the Obj-C
header, and in the future also .swiftinterface etc.)

Despite the as-of-this-change removed comment claiming this is because
the Swift compiler does not support outputting .swiftmodule and
object files in one invocation, this was seemingly never the case,
however we still keep a separate target for the object files here
since generation of the interface alone is faster, which potentially
increases parallelization when building large module trees.

See <https://forums.swift.org/t/generating-export-header-from-swiftmodule-breaking-change-in-swift-6-3/85817>.

dblsaiko committed:
swift: Track extra output files from swiftc

dblsaiko committed:
Check whether Swift/Obj-C interop is available

dblsaiko commented:
The big thing I still don't like about this is the way the C export header is handled, and also how the Swift->Swift interface has been handled all this time. Right now, the C export header gets implicitly added as a ninja dependency (see determine_swift_dep_headers), but you still have to explicitly add the private target include dir to the dependent target to actually be able to import it.

I have two ideas for this:

1. Automatically add the include directory to C targets linking against a Swift target. This matches the behavior of Swift -> Swift target dependencies where you only need to link it to be able to import it.
2. Stop automatically adding dependencies to the Swift interface and the C header. Instead, have target types expose a dependency object for the interface that you have to add to your target.

Conceptually I prefer 2., but it would be a breaking change for Swift support. It matches more closely to what is going on internally, which is just like the C model, and allows you to control linking vs. includes more correctly. For example, think setting up a bunch of static libraries linked into one shared library, where the interfaces are kept separate; this is pretty much impossible to do right now unlike with C because you cannot make the interface available to a target without also linking the library.

WDYT?

jpakkane commented:
> Conceptually I prefer 2., but it would be a breaking change for Swift support.

We don't have _that_ many Swift users as far as I know so breaking things should be possible. Especially if it makes things better/more maintainable in the long run.

dblsaiko committed:
swift: Hotfix for compiler.compiles()
